packages/opentelemetry-context-base/src/types.ts

```diff
@@ -0,0 +1,51 @@
+/**
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** Default type for functions */
+// tslint:disable:no-any
+export type Func<T> = (...args: any[]) => T;
+
+/** A key/value object for anything that can be stored in the context */
+export interface ContextStore {
+  [key: string]: unknown;
+}
+
+export interface ContextManager {
+  /**
+   * Binds the trace context to the given function.
+   * This is necessary in order to create child spans correctly in functions
+   * that are called asynchronously (for example, in a network response
+   * handler).
+   * @param fn A function to which to bind the trace context.
+   */
+  wrapFunction<T>(fn: Func<T>): Func<T>;
+
+  /**
+   * Binds the trace context to the given event emitter.
+   * This is necessary in order to create child spans correctly in event
+   * handlers.
+   * @param emitter An event emitter whose handlers should have
+   *     the trace context binded to them.
+   */
+  wrapEmitter(emitter: NodeJS.EventEmitter): void;
```

Comment:
I believe the `wrapEmitter` is not needed anymore since we target node 8, `async_hooks` propagate the context correctly with event emitters.

Also it might be a problem to compile to the browser since the EventEmitter will not be available in this environment.

I added it so we can make a decision on its utility

Reply:
Thanks for thinking of the browser here! If we can remove it, that would be awesome.

When I built OpenCensus Web, I had to do a workaround to avoid having to bring in the Node TS types (which felt like an extra dependency and risks type conflicts). So what I ended up doing was writing a script that did some auto-replacements and I also inlined the type for the NodeJS.EventEmitter.

If we do keep this, I would ask us to consider inlining the event emitter type in this library so that browser users of it don't need the Node TS types to make it compile for TS.

Reply:
`async_hooks` doesn't propagate the correct context, at least to my knowledge. It will propagate the ID where `emit` is run and not where `on` is run.

Also, it would make for a simpler interface to have a single `wrap` method that supports multiple types. It will also allow extensions without breaking the contract.

Reply:
Yeah i though of using only one method in the first place but i remove it later thinking it was bringing too much complexity. Thinking it again it might be problematic with the browser API once again, i will change to only one method